transport-classes-io_uring/src/main/java/io/netty/channel/uring/AbstractIOUringChannel.java

```diff
@@ -0,0 +1,1111 @@
+                    }
+                    fulfillConnectPromise(connectPromise, wasActive);
+                } catch (Throwable t) {
+                    fulfillConnectPromise(connectPromise, annotateConnectException(t, requestedRemoteAddress));
+                } finally {
+                    if (!connectStillInProgress) {
+                        // Check for null as the connectTimeoutFuture is only created if a connectTimeoutMillis > 0
+                        // is used
+                        // See https://github.com/netty/netty/issues/1770
+                        cancelConnectTimeoutFuture();
+                        connectPromise = null;
+                    } else {
+                        // The connect was not done yet, register for POLLOUT again
+                        schedulePollOut();
+                    }
+                }
+            } else if (!socket.isOutputShutdown()) {
+                // Try writing again
+                super.flush0();
+            }
+        }
+
+        /**
+         * Called once a write was completed.
+         *
+         * @param res   the result.
+         * @param flags the flags.
+         * @param data  the data that was passed when submitting the op.
+         */
+        final void writeComplete(int res, int flags, short data) {
+            if ((ioState & CONNECT_SCHEDULED) != 0) {
+                // The writeComplete(...) callback was called because of a sendmsg(...) result that was used for
+                // TCP_FASTOPEN_CONNECT.
+                freeMsgHdrArray();
+                if (res > 0) {
+                    // Connect complete!
+                    outboundBuffer().removeBytes(res);
+
+                    // Explicit pass in 0 as this is returned by a connect(...) call when it was successful.
+                    connectComplete(0, flags, data);
+                } else if (res == ERRNO_EINPROGRESS_NEGATIVE || res == 0) {
+                    // This happens when we (as a client) have no pre-existing cookie for doing a fast-open connection.
+                    // In this case, our TCP connection will be established normally, but no data was transmitted at
+                    // this time. We'll just transmit the data with normal writes later.
+                    // Let's submit a normal connect.
+                    submitConnect((InetSocketAddress) requestedRemoteAddress);
+                } else {
+                    // There was an error, handle it as a normal connect error.
+                    connectComplete(res, flags, data);
+                }
+                return;
+            }
+            assert numOutstandingWrites > 0;
+            --numOutstandingWrites;
+
+            boolean writtenAll = writeComplete0(res, flags, data, numOutstandingWrites);
+            if (!writtenAll && (ioState & POLL_OUT_SCHEDULED) == 0) {
+                // We were not able to write everything, let's register for POLLOUT
+                schedulePollOut();
```

Comment:
We already use blocking writes so `writtenAll` should always return true in our case. I only kept it to ensure it would work with non-blocking writes as well.